modules/core/src/main/java/org/apache/ignite/internal/processors/query/aware/IndexRebuildStateStorage.java

```diff
@@ -0,0 +1,412 @@
+     */
+    public void onQueryStart() {
+        ctx.internalSubscriptionProcessor().registerMetastorageListener(this);
+    }
+
+    /**
+     * Callback on start of {@link GridCacheProcessor}.
+     */
+    public void onCacheKernalStart() {
+        lock.lock();
+
+        try {
+            if (!states.isEmpty()) {
+                Set<String> toComplete = new HashSet<>(states.keySet());
+                toComplete.removeAll(ctx.cache().cacheDescriptors().keySet());
+
+                for (String cacheName : toComplete)
+                    completeRebuildIndexes(cacheName);
+            }
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Callback on kernel stop.
+     */
+    public void onKernalStop() {
+        lock.lock();
+
+        try {
+            stopGrid = true;
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Callback on start of rebuild cache indexes.
+     * <p/>
+     * Adding an entry that rebuilding the cache indexes in progress.
+     * If the cache is persistent, then add this entry to the MetaStorage.
+     *
+     * @param cacheCtx Cache context.
+     * @see #onFinishRebuildIndexes
+     */
+    public void onStartRebuildIndexes(GridCacheContext cacheCtx) {
+        lock.lock();
+
+        try {
+            assert !stopGrid;
+
+            IndexRebuildState newState = of(cacheCtx);
+
+            if (!newState.saved())
+                states.putIfAbsent(newState.cacheName(), newState);
+            else {
+                toRmv.remove(newState.cacheName());
+
+                states.compute(newState.cacheName(), (k, preState) -> {
+                    if (preState == null)
+                        return newState;
+                    else
+                        return preState.completed(false);
+                });
+
+                metaStorageOperation(metaStorage -> {
+                    assert metaStorage != null;
+
+                    metaStorage.write(metaStorageKey(newState), of(newState));
+                });
+            }
+        }
+        finally {
+            lock.unlock();
+        }
+    }
+
+    /**
+     * Callback on finish of rebuild cache indexes.
+     * {@link #onStartRebuildIndexes} is expected to have been called before.
+     * <p/>
+     * If the cache is persistent, then we mark that the rebuilding of the
+     * indexes is completed and the entry will be deleted from the MetaStorage
+     * at the end of the checkpoint. Otherwise, delete the index rebuild entry.
+     *
+     * @param cacheCtx Cache context.
+     * @see #onStartRebuildIndexes
+     */
+    public void onFinishRebuildIndexes(GridCacheContext cacheCtx) {
+        lock.lock();
+
+        try {
+            String cacheName = cacheCtx.name();
+
+            IndexRebuildState state = states.get(cacheName);
+
+            assert state != null : cacheName;
+
+            if (state.saved())
+                state.completed(true);
```

Comment:
leakage, i hope ? why u don`t call _states.remove_ here ?

Reply:
For in-memory cache, we can immediately delete, for persistent only after the end of the checkpoint.